lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb

```diff
@@ -0,0 +1,202 @@
+
+          # Get all tasks
+          tasks, refresh_state_uuid = {}, SecureRandom.uuid
+          params.to_a.each do |task|
+            if task['task_id'].blank? || task['source_ref'].blank?
+              logger.warn("ServiceInstance#refresh - missing data for task: #{task}")
+              next
+            end
+            tasks[task['task_id']] = task['source_ref']
+
+            if tasks.length == REFS_PER_REQUEST_LIMIT
+              refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid)
+              tasks = {}
+            end
+          end
+
+          refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid) unless tasks.empty?
+        rescue => err
+          logger.error("ServiceInstance#refresh - Error: #{err.message}\n#{err.backtrace.join("\n")}")
+        end
+
+        private
+
+        attr_accessor :params, :source_id
+
+        def required_params
+          %w[source_id source params]
+        end
+
+        def params_missing?
+          is_missing = false
+          required_params.each do |attr|
+            if (is_missing = send(attr).blank?)
+              logger.error("ServiceInstance#refresh - Missing #{attr} for the availability_check request [Source ID: #{source_id}]")
+              break
+            end
+          end
+
+          is_missing
+        end
+
+        def refresh_part(tasks, refresh_state_uuid, refresh_state_part_uuid)
+          tasks_id = tasks.keys.join(' | id: ')
+
+          # API request, nodes and jobs under workflow job not needed
+          query_params = {
+            :id__in    => tasks.values.join(','),
+            :page_size => limits['service_instances']
+          }
+
+          if on_premise?
+            refresh_part_on_premise(tasks_id, refresh_state_uuid, query_params)
+          else
+            refresh_part_cloud(tasks_id, refresh_state_uuid, refresh_state_part_uuid, query_params)
+          end
+        end
+
+        def refresh_part_on_premise(_tasks_id, refresh_state_uuid, query_params)
+          refresh_state_started_at = Time.now.utc
+          receptor_params = {:accept_encoding => 'gzip', :fetch_all_pages => true}
+
+          receiver = TopologicalInventory::AnsibleTower::Receptor::AsyncReceiver.new(self, connection,
+                                                                                     'service_instances',
+                                                                                     refresh_state_uuid,
+                                                                                     refresh_state_started_at,
+                                                                                     :sweeping_enabled => false)
+          get_service_instances(connection, query_params,
+                                :on_premise        => true,
+                                :receptor_receiver => receiver,
+                                :receptor_params   => receptor_params)
+        end
+
+        public
```

Comment:
I think this method should be moved up (removing the public specifier) so we can just stick to one `private` definition rather than switching back and forth.

Reply:
I wanted to group related methods together, but it seems it's not a good idea